File: 383-ransom-note/ransom-note.ts
function canConstruct(ransomNote: string, magazine: string): boolean {
    const map: Map<string, number> = new Map()

    for (let char of magazine) {
        map.set(char, (map.get(char) || 0) + 1)
    }

    for (let char of ransomNote) {
        if (!map.has(char)) return false
        else {
            map.set(char, (map.get(char) || 0) - 1)
            if (map.get(char) === 0) {
                map.delete(char)
            }
        }
    }

    return true
};

File: 383-ransom-note/ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        magazine_counts = defaultdict(int)

        for char in magazine:
            magazine_counts[char] += 1

        for char in ransomNote:
            if magazine_counts[char] == 0:
                return False
            
            magazine_counts[char] -= 1

        return True